File: go-backend-interview-prep/11_system_design_in_go/05_observability_basics_mini/main.py
"""Observability basics -- Python equivalent with request ID, logging, metrics."""

import json
import os
import random
import string
import threading
import time
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.request import urlopen


# --- Request ID ---

def generate_request_id():
    return "".join(random.choices(string.hexdigits[:16], k=16))


# --- Metrics ---

class Metrics:
    def __init__(self):
        self.lock = threading.Lock()
        self.total_requests = 0
        self.total_errors = 0
        self.total_latency_ms = 0.0

    def record(self, status_code, latency_ms):
        with self.lock:
            self.total_requests += 1
            self.total_latency_ms += latency_ms
            if status_code >= 400:
                self.total_errors += 1

    def snapshot(self):
        with self.lock:
            avg = 0.0
            if self.total_requests > 0:
                avg = self.total_latency_ms / self.total_requests
            return {
                "requests_total": self.total_requests,
                "errors_total": self.total_errors,
                "latency_total_ms": round(self.total_latency_ms, 2),
                "latency_avg_ms": round(avg, 2),
            }


metrics = Metrics()


# --- Handler ---

class Handler(BaseHTTPRequestHandler):
    def do_GET(self):
        start = time.time()
        request_id = generate_request_id()

        if self.path == "/hello":
            body = {"message": "hello", "request_id": request_id}
            status = 200
        elif self.path == "/error":
            body = {"error": "something went wrong"}
            status = 500
        elif self.path == "/metrics":
            snap = metrics.snapshot()
            text = "\n".join(f"{k} {v}" for k, v in snap.items())
            self.send_response(200)
            self.send_header("Content-Type", "text/plain")
            self.end_headers()
            self.wfile.write(text.encode())
            return
        else:
            self.send_response(404)
            self.end_headers()
            return

        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("X-Request-ID", request_id)
        self.end_headers()
        self.wfile.write(json.dumps(body).encode())

        latency_ms = (time.time() - start) * 1000
        metrics.record(status, latency_ms)

        # Structured log
        print(f"request_id={request_id} method={self.command} "
              f"path={self.path} status={status} latency={latency_ms:.1f}ms")

    def log_message(self, format, *args):
        pass


# --- Demo ---

def run_demo():
    time.sleep(0.5)

    print("--- sending requests ---\n")
    for i in range(5):
        resp = urlopen("http://localhost:9005/hello", timeout=2)
        body = json.loads(resp.read())
        print(f"GET /hello -> request_id={body['request_id']}")

    resp = urlopen("http://localhost:9005/error", timeout=2)
    print(f"GET /error -> status={resp.status}")

    print("\n--- GET /metrics ---\n")
    resp = urlopen("http://localhost:9005/metrics", timeout=2)
    print(resp.read().decode())

    print("demo done")
    os._exit(0)


if __name__ == "__main__":
    print("observability demo on :9005")
    t = threading.Thread(target=run_demo, daemon=True)
    t.start()
    server = HTTPServer(("", 9005), Handler)
    server.serve_forever()

File: go-backend-interview-prep/11_system_design_in_go/05_observability_basics_mini/main.go
package main

import (
	"context"
	"crypto/rand"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"sync/atomic"
	"time"
)

// --- Request ID ---

type ctxKey string

const requestIDKey ctxKey = "request_id"

func generateRequestID() string {
	b := make([]byte, 8)
	rand.Read(b)
	return fmt.Sprintf("%x", b)
}

func getRequestID(ctx context.Context) string {
	if id, ok := ctx.Value(requestIDKey).(string); ok {
		return id
	}
	return "unknown"
}

// --- Metrics ---

type Metrics struct {
	totalRequests atomic.Int64
	totalErrors   atomic.Int64
	totalLatencyMs atomic.Int64
}

var metrics Metrics

// --- Middleware: Request ID ---

func requestIDMiddleware(next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id := generateRequestID()
		ctx := context.WithValue(r.Context(), requestIDKey, id)
		w.Header().Set("X-Request-ID", id)
		next(w, r.WithContext(ctx))
	}
}

// --- Middleware: Logging + Metrics ---

type statusRecorder struct {
	http.ResponseWriter
	statusCode int
}

func (sr *statusRecorder) WriteHeader(code int) {
	sr.statusCode = code
	sr.ResponseWriter.WriteHeader(code)
}

func loggingMiddleware(next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()
		recorder := &statusRecorder{ResponseWriter: w, statusCode: 200}

		next(recorder, r)

		latency := time.Since(start)
		reqID := getRequestID(r.Context())

		// Update metrics
		metrics.totalRequests.Add(1)
		metrics.totalLatencyMs.Add(latency.Milliseconds())
		if recorder.statusCode >= 400 {
			metrics.totalErrors.Add(1)
		}

		// Structured log line
		log.Printf("request_id=%s method=%s path=%s status=%d latency=%v",
			reqID, r.Method, r.URL.Path, recorder.statusCode, latency)
	}
}

// --- Handlers ---

func helloHandler(w http.ResponseWriter, r *http.Request) {
	reqID := getRequestID(r.Context())
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]string{
		"message":    "hello",
		"request_id": reqID,
	})
}

func errorHandler(w http.ResponseWriter, r *http.Request) {
	w.WriteHeader(http.StatusInternalServerError)
	json.NewEncoder(w).Encode(map[string]string{
		"error": "something went wrong",
	})
}

func metricsHandler(w http.ResponseWriter, r *http.Request) {
	total := metrics.totalRequests.Load()
	errors := metrics.totalErrors.Load()
	latency := metrics.totalLatencyMs.Load()

	avgLatency := float64(0)
	if total > 0 {
		avgLatency = float64(latency) / float64(total)
	}

	w.Header().Set("Content-Type", "text/plain")
	fmt.Fprintf(w, "requests_total %d\n", total)
	fmt.Fprintf(w, "errors_total %d\n", errors)
	fmt.Fprintf(w, "latency_total_ms %d\n", latency)
	fmt.Fprintf(w, "latency_avg_ms %.2f\n", avgLatency)
}

// --- Chain helper ---

func chain(handler http.HandlerFunc, middlewares ...func(http.HandlerFunc) http.HandlerFunc) http.HandlerFunc {
	for i := len(middlewares) - 1; i >= 0; i-- {
		handler = middlewares[i](handler)
	}
	return handler
}

// --- Demo ---

func main() {
	mux := http.NewServeMux()

	// Apply middleware chain
	mux.HandleFunc("/hello", chain(helloHandler, requestIDMiddleware, loggingMiddleware))
	mux.HandleFunc("/error", chain(errorHandler, requestIDMiddleware, loggingMiddleware))
	mux.HandleFunc("/metrics", metricsHandler)

	// Demo: simulate requests
	go func() {
		time.Sleep(500 * time.Millisecond)
		client := &http.Client{Timeout: 2 * time.Second}

		fmt.Print("--- sending requests ---\n\n")

		// Normal requests
		for i := 0; i < 5; i++ {
			resp, _ := client.Get("http://localhost:9005/hello")
			var body map[string]string
			json.NewDecoder(resp.Body).Decode(&body)
			resp.Body.Close()
			fmt.Printf("GET /hello -> request_id=%s\n", body["request_id"])
		}

		// Error request
		resp, _ := client.Get("http://localhost:9005/error")
		fmt.Printf("GET /error -> status=%d\n", resp.StatusCode)
		resp.Body.Close()

		// Check metrics
		fmt.Print("\n--- GET /metrics ---\n\n")
		resp, _ = client.Get("http://localhost:9005/metrics")
		buf := make([]byte, 1024)
		n, _ := resp.Body.Read(buf)
		resp.Body.Close()
		fmt.Println(string(buf[:n]))

		log.Println("demo done")
		go func() { time.Sleep(100 * time.Millisecond); log.Fatal("exit") }()
	}()

	fmt.Println("observability demo on :9005")
	log.Fatal(http.ListenAndServe(":9005", mux))
}
